### DISCUSSION TITLE
How to get the progress bar while transcribing?

### DISCUSSION THREAD

**Discussion by benjieperez** [openai/whisper#850] [Show and tell]
Here is my code.

import whisper

# Load Whisper model Large
MODEL = whisper.load_model("model/medium.pt", in_memory=True)
# Set options
translate_options = dict(task="translate", **dict(language=translate_language, beam_size=5, best_of=5))
result_transcribe = MODEL.transcribe(audio, **transcribe_options, fp16=False, verbose=False)

When i enable verbose=False, this is the terminal output while transcribing, 

Detected language: Turkish
60%|███████████████████████████████████████████████████████████████████████████        | 4058/6058 [00:21<00:00, 277.58frames/s]
100%|███████████████████████████████████████████████████████████████████████████        | 6058/6058 [00:21<00:00, 277.58frames/s]

I want to get the progress bar while its transcribing is it possible?

Thanks,

**Comment by mayeaux** [comment]
This is how I do it in JavaScript: https://github.com/mayeaux/generate-subtitles/blob/master/helpers/formatStdErr.js#L9

**Reply by benjieperez** [reply]
Yeah this is what I exactly transposed to python but I don't know how to get those logs from the python library itself.

**Reply by glangford** [reply]
The code in `transcribe.py` says
https://github.com/openai/whisper/blob/02aa851a4910201f0db56960064d7e121a01002c/whisper/transcribe.py#L170

I get the progress bar on my platform with a call to `transcribe()` as long as I have `verbose=False`. 

`whisper` uses `tqdm` for the progress bar, I wonder if you are running into an issue for your platform or environment. See for example
[TQDM FAQ and Known Issues](https://github.com/tqdm/tqdm#faq-and-known-issues)

You can call `tqdm` from the shell if you want to test that it is working independent of `whisper`.

**Reply by mayeaux** [reply]
> Yeah this is what I exactly transposed to python but I don't know how to get those logs from the python library itself.

I run it as a node process and just read the output. I found getting the realtime data from Python tricky but you can add functionality where `tqdm` is called

**Reply by benjieperez** [reply]
Yes, I've read also Line 170, but it shows that the function transcribe only returns the text, segments, and language. Not the progress bar, it was an output as the system log. Do you have any ideas how to extract it? I'm using FastAPI and cant get it through out of it.

**Reply by benjieperez** [reply]
# I get the progress bar on my platform with a call to transcribe() as long as I have verbose=False

How did you do it? because when i printed out the MODEL.transcribe(verbose=False), it only returns segments, text, language

**Reply by glangford** [reply]
I don't know FastAPI, but as you say
> ...the function `transcribe` only returns the text, segments, and language. Not the progress bar, **it was an output as the system log**. 

Right, `transcribe` only *returns* this data at the end (line 246), and uses `tqdm` **before it returns** to display progress as console output.

`tqdm` is writing periodically to `stderr`. Note also that @mayeaux 's code sample reads `stderr` to parse the output from `tqdm`.

**Reply by benjieperez** [reply]
I try to replicate what @mayeaux has done using Node Process.

``
    def transcribe(command_line):
	    try:
		    command_args = command_line.split(' ')
		    print("Running Command:",command_line)
		    # start whisper process
		    process = subprocess.Popen(command_args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    
		    while True:
			    print(process.stdout.readline())
			    print(process.stderr.readline())
	    except BaseException as e:
		    print('Error on line {}'.format(sys.exc_info()[-1].tb_lineno), type(e).__name__, e)
		    pass
``


This was the output from the function.
b'Detected language: Turkish\n'
b''
b''
b''
b''
b''
b''
The output i get is only the stdout, and when the process finish no progress bar is showing, cant track the progress while processing. Any good suggestion or am I missing something? Thank you.

**Reply by glangford** [reply]
> The output i get is only the stdout, and when the process finish no progress bar is showing, cant track the progress while processing.

I tried that with the sample code and got the same thing. :) `readline()` is blocking so you probably don't want to be reading both `stdout` and `stderr` in the loop in that way.

The issue is likely that `tqdm` is not writing newline terminated lines (they are terminated with \r).

To get it to work, try changing your `Popen` to 
```Python
process = subprocess.Popen(
            command_args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
```
Then reading `stderr` seems to work. Hopefully that helps.

**Reply by glangford** [reply]
eg.
```Python
process = subprocess.Popen(
            command_args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)

        while True:
            line = process.stderr.readline()
            if line == "":
                break
            print(line)
```

**Reply by benjieperez** [reply]
After trying the snippet, yes its now printing the progress bar, it doesn't print at real time. The progress bar will appear when the process is finished not every single time the progress bar is being updated.


Date Run: 2023-01-19 10:36
Running Command: whisper files_uploaded/french_news_audio.mp3 --model=medium --model_dir=model/ --verbose=False --fp16=False --task=transcribe --output_dir=output/
STDOUT: Detected language: French

This progress bar bellow will not print until the process is done.
  0%|          | 0/9464 [00:00<?, ?frames/s]

 22%|██▏       | 2102/9464 [00:11<00:38, 189.05frames/s]

 46%|████▌     | 4320/9464 [00:21<00:25, 202.01frames/s]

 64%|██████▍   | 6072/9464 [00:30<00:16, 201.55frames/s]

 78%|███████▊  | 7360/9464 [00:36<00:10, 204.65frames/s]

 78%|███████▊  | 7360/9464 [00:49<00:10, 204.65frames/s]

 96%|█████████▋| 9116/9464 [01:10<00:03, 96.89frames/s] 

 96%|█████████▋| 9116/9464 [01:29<00:03, 96.89frames/s]

100%|██████████| 9464/9464 [01:40<00:00, 57.47frames/s]

100%|██████████| 9464/9464 [01:40<00:00, 93.98frames/s]

Total Time Processed: --- 116.53934979438782 seconds ---

**Reply by jongwook** [reply]
@benjieperez Have you tried setting the `PYTHONUNBUFFERED` environment variable to `1` (or anything other than `0`)?

**Reply by benjieperez** [reply]
How to set PYTHONUNBUFFERED? No idea on this.

**Reply by glangford** [reply]
@benjieperez 
* https://github.com/docker-library/python/issues/604#issue-860314217
* https://docs.python.org/3/library/sys.html#sys.stdout
* https://stackoverflow.com/questions/2231227/python-subprocess-popen-with-a-modified-environment

**Reply by glangford** [reply]
@benjieperez  ...something along these lines is worth a try
```Python
import subprocess, os
whisper_env = os.environ.copy()
whisper_env["PYTHONUNBUFFERED"] = "1"
subprocess.Popen( ... , env=whisper_env)
```

**Reply by benjieperez** [reply]
> @benjieperez ...something along these lines is worth a try
> 
> ```python
> import subprocess, os
> whisper_env = os.environ.copy()
> whisper_env["PYTHONUNBUFFERED"] = "1"
> subprocess.Popen( ... , env=whisper_env)
> ```

Still no luck, the progress bar shows after the process not while processing the audio.

**Reply by benjieperez** [reply]
------------------------------------
Date Run: 2023-01-23 10:44
Running Command: whisper files_uploaded/63c8cb48abc63045d80fd5d2.wav --model=medium --model_dir=model/ --verbose=False --fp16=False --task=transcribe --output_dir=output/63c8cb48abc63045d80fd5d2
STDOUT: Detected language: Turkish

STDOUT: 
STERR: 

None
(This is blank, after the process finished, this will be the result)
STERR:   0%|          | 0/6060 [00:00<?, ?frames/s]

{'percentDone': '0%', 'timeElapsed': '00:00', 'speed': 0, 'percentDoneAsNumber': 0.0, 'timeRemaining': {'string': '00:00', 'hoursRemaining': 0, 'minutesRemaining': 0, 'secondsRemaining': 0}}
STERR:  36%|███▋      | 2200/6060 [00:10<00:17, 215.65frames/s]

{'percentDone': '36%', 'timeElapsed': '00:10', 'speed': '215.65', 'percentDoneAsNumber': 36.0, 'timeRemaining': {'string': '00:17', 'hoursRemaining': 0, 'minutesRemaining': 0, 'secondsRemaining': 17}}
STERR:  74%|███████▍  | 4500/6060 [00:18<00:06, 256.08frames/s]

{'percentDone': '74%', 'timeElapsed': '00:18', 'speed': '256.08', 'percentDoneAsNumber': 74.0, 'timeRemaining': {'string': '00:06', 'hoursRemaining': 0, 'minutesRemaining': 0, 'secondsRemaining': 6}}
STERR: 100%|██████████| 6060/6060 [00:21<00:00, 295.05frames/s]

{'percentDone': '100%', 'timeElapsed': '00:21', 'speed': '295.05', 'percentDoneAsNumber': 100.0, 'timeRemaining': {'string': '00:00', 'hoursRemaining': 0, 'minutesRemaining': 0, 'secondsRemaining': 0}}
STERR: 100%|██████████| 6060/6060 [00:21<00:00, 276.00frames/s]

{'percentDone': '100%', 'timeElapsed': '00:21', 'speed': '276.00', 'percentDoneAsNumber': 100.0, 'timeRemaining': {'string': '00:00', 'hoursRemaining': 0, 'minutesRemaining': 0, 'secondsRemaining': 0}}
STERR: 
Total Time Processed: --- 43.48728823661804 seconds ---

**Comment by zefr0x** [comment]
I think it would be great to have a function like `whisper.transcribe()`, that yields results rather then waiting to the end then returning every thing at ones.

I think it's not complicated, since `whisper.transcribe()` can print result in realtime when we set `verbose=True`.

**Comment by aadnk** [comment]
Another option is to override the `tqdm.tqdm` progress bar in Whisper itself:
```python
import os
import sys
import tqdm
import urllib.request 

class _CustomProgressBar(tqdm.tqdm):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._current = self.n  # Set the initial value
        
    def update(self, n):
        super().update(n)
        self._current += n
        
        # Handle progress here
        print("Progress: " + str(self._current) + "/" + str(self.total))

# Inject into tqdm.tqdm of Whisper, so we can see progress
import whisper.transcribe 
transcribe_module = sys.modules['whisper.transcribe']
transcribe_module.tqdm.tqdm = _CustomProgressBar

import whisper
model = whisper.load_model("medium")

if not os.path.exists("sample1.wav"):
    urllib.request.urlretrieve("https://github.com/itsupera/audiobook_alignment/raw/main/samples/sample1.wav", "sample1.wav")

result = model.transcribe("sample1.wav", language="Japanese", fp16=False, verbose=None)
print(result['text'])
```
That way, you don't have launch Whisper in another process, or parse the output. I have a complete example of how this can be done here:
* https://huggingface.co/spaces/aadnk/whisper-webui/blob/main/src/hooks/whisperProgressHook.py#L95

I use it to pass progress from Whisper into Gradio. But yeah, hopefully there will be a proper API soon so we don't have to rely on method hooks or parsing the console output.

**Reply by bees4ever** [reply]
love to see it as a standard callback of the transcribe function

**Reply by J-Klinke** [reply]
I hope I'll get a reply on this old discussion, but I having trouble with your example. I tried with this code snippet and the code from the linked repo, but the output is still buffered / appears as transcribe() terminates. Can you think of anything I'm doing wrong?

For my test i copied your code from whisperProgressHook.py and progressListener.py into one python script and only adapted the audio file and the language.
Thanks a lot!

**Comment by jbeltran73-2** [comment]
```python
#!/usr/bin/env python3

import whisper

model = whisper.load_model("large-v3")
result = model.transcribe("scale.m4a")

with open("transcription.txt", "w") as f:
        f.write(result["text"])
```
I don't see the progress bar with this simple code. I've searched and transcribe.py has tqdm already. 
What I could be doing wrong? 

Thank you

**Reply by dkleptsov** [reply]
Many functions have "verbose" boolean parameter. It controls how much information function will print to terminal.

Transcribe also has this parameter:
result = model.transcribe("scale.m4a") - will print nothing
result = model.transcribe("scale.m4a", verbose=False) - will print only progress bar
result = model.transcribe("scale.m4a", verbose=True) - will print detailed information

**Reply by jbeltran73-2** [reply]
Thank you trying it.

**Comment by hdino** [comment]
PR #2625 adds a progress callback to the `transcribe` function. It’d be nice to have it merged.
